### ISSUE TITLE
Proposition:  use goog.module

### ISSUE THREAD

**Issue by gberaudo** [openlayers/openlayers#5226]
I have been experimenting with goog.module. The results look promising; before investing more time I would be pleased to get your feedback.

In relation to #4128.

Rationnal: gradual and non-breaking step towards ES6

Details:

OL3 dependencies are handled using goog.provide / goog.require.
The symbols are evaluated in the global scope and made available in a hierarchy attached to the global object.

On the other hand, ES6 and AMD/common.js evalutate the code in an isolated scope.
It is not attached to the global object.

goog.module goals is to allow both styles, at the same time, making it possible to transition gradually towards ES6 support, without breaking backward compatibility.

A good explanation about goog.modules is here:
https://github.com/google/closure-library/wiki/goog.module:-an-ES6-module-like-alternative-to-goog.provide

With @fredj, we created a branch in closure-util:
https://github.com/gberaudo/closure-util/tree/goog_module
It makes goog.module understood by closure-util and do the wrapping to isolate the module.

Here are experiments in OL3 code:
https://github.com/gberaudo/ol3/tree/goog_modules
It shows how to migrate namespaces, constructors, typedefs and enums.

Notes:
A file using goog.module can have only one goog.module and no goog.provide.
Typedefs must be moved to their own files but keep goog.provide.
There are some issues with jsdoc parsing for generating the exports (see commits).

**Comment by probins** [comment]
ISTM there's 2 types of task here:

1. those, like moving the typedefs, which don't affect the actual code and can be done anyway; I'd suggest putting these in a separate directory, like the externs. Also in this category are removing the `goog.require`s which aren't needed anyway (as in #5210); and renaming files to match module names, such as `layer/vectorlayer.js` to `layer/Vector.js`
2. changing the actual code, as with changing the current global vars to the `exports` var. The problem with fundamental changes like this is that they're likely to quickly lead to merge conflicts, so if this can be done on a file by file basis, that would be better.

I was going to offer to put some time into the first, if people agree that makes sense, so happy to work with you on this if that can be arranged.

I'd agree that using `goog.module` could be a good transitional step. It doesn't remove dependency on GCC, but it does move in that direction by changing to something that is module-ish.

**Comment by tschaub** [comment]
Looks promising @gberaudo.  I agree with @probins about pulling out the changes that aren't specific to `goog.module` - assuming these are changes we can make without upgrading `closure-util`.

The main motivation behind #4128 is to make OpenLayers work with mainstream JavaScript tools.  So this is definitely not a solution for that.  But it seems like it could make it possible to transpile to ES6 or CommonJS modules for a release (assuming there were no dependencies on the Closure Library).  And this would be a big win (to have a release that let people create custom builds with the Compiler or use the module bundler of their choice).

So as long as we don't stop the effort to get rid of our dependency on the Closure Library, I'm in support of this work.

**Comment by probins** [comment]
> it seems like it could make it possible to transpile to ES6 or CommonJS modules

it's very similar to CJS. On the face of it, you could probably even convert with a string replace. The `goog.module` isn't needed and can be discarded; `goog.require` -> `require`; and `xx.yy.zz` -> `./xx/yy/zz.js` (assuming there's a 1-1 file-module relationship). And I think that's it.

**Comment by gberaudo** [comment]
@probins, @tschaub, thanks for the feedbacks.
@probins, thank you for proposing to help, it is gladly welcomed!

In the moving task, in addition to `typedef`s, each `constructor` or `enum` must be moved to its own file in order to reach one `goog.provide` per file. Regarding typedefs, I have no definitive opinion; it is more intuitive to have one file per typedef, on the other if it can simplify stuff and actually work, I am fine with a single file typedef like discussed in https://github.com/openlayers/ol3/pull/5210#issuecomment-209586078.

Both the moving task and the module task have big impact on the code that could lead to merge conflicts so changes should be done incrementally, with PRs on a file per file or on a small set of files.

Once the way to handle typedef is decided, we can start the moving task.
Once changes land in closure-util, and the last issues about exports and documentation are figured out we can start the module task on already moved files.

The module task will actually be split in two steps:
- first, using goog.module in all files (except typedefs);
- then, using `goog.require` with assignements.

Does it make sense for everyone?

**Comment by probins** [comment]
> changes should be done incrementally, with PRs on a file per file or on a small set of files

there's still going to be merge problems, for example with outstanding PRs. It might be a good idea to have a feature freeze, clear out the current PRs and whatever's still outstanding for removing Closure lib, before tackling all this moving about of code.

**Comment by probins** [comment]
> The module task will actually be split in two steps:
> 
> first, using goog.module in all files (except typedefs);
> then, using goog.require with assignements.

to make sure I understand this:

* `goog.module(namespace)` replaces `goog.provide(namespace)`; neither of these has an equivalent in either ES6 or CJS, which don't use namespaces
* `goog.require(namespace)` within a `goog.module` can be assigned to an module var, which can then be used within the module; this is the same as CJS except that it uses a namespace instead of a path/url
* `exports` is similar to CJS
* `declareLegacyNamespace()` can be used so that other files which still use `goog.provide(namespace)` can use the namespace as before without having to mess about with `goog.scope` as described [on the Compiler wiki](https://github.com/google/closure-library/wiki/goog.module:-an-ES6-module-like-alternative-to-goog.provide#how-do-i-use-a-googmodule-from-a-traditional-closure-file). Once all files have been changed to modules with exports, then `declareLegacyNamespace()` can be removed.

Have I got that right? Why, in your branch, are you using `goog.require()` and not `var x = goog.require()? Is this also for bc?

**Comment by gberaudo** [comment]
The result of `goog.require('something')` may only be put in a variable if `something` is provided by a `goog.module('something')`. If the symbol is provided by a `goog.provide('something')` it can not be put in a variable.

`goog.module.declareLegacyNamespace()` is also useful for the applications and libraries using OL3. They will be able to:
- stay unchanged: backward compatibility;
- directly use `goog.module`/`goog.require`: new applications;
- a mix of the two: migrating applications.
As a consequence, we would keep the call for as long as OL3 uses `goog.modules` in order to give the longest possible time to users for migrating.

In my branch I could start using `var x = goog.require()` in modules which depends on modules already migrated.

Does it clarify?

**Comment by probins** [comment]
right, so `goog.require(namespace)` establishes a dependency; it is not an import. Whereas `var x = goog.require(module)` is an import, and is the equivalent of CJS require and ES6 import.

**Comment by probins** [comment]
Do all enums have to be in their own module? Take `ol.extent` for example. This has 2 enums, but do they really each need their own module? If I were writing this myself as a standard module, there's no way I would put a simple object like this in its module.

**Comment by gberaudo** [comment]
If it works by having them as properties of the `ol.extent` that is fine to me.
- [ ] check if enums can be properties of a namespace and still be used in annotations.

**Comment by probins** [comment]
another question, on typedefs. In your branch, `src/ol/attributionlike.js` has a `goog.provide`, but this isn't present in the existing code. So do typedefs need a `goog.provide` at all? It's not something I've paid much attention to, but there seems to be quite a bit of inconsistency in the current code on whether `goog.provide/require` are present or not.

**Comment by probins** [comment]
`ol.events` is another example. This has 3 typedefs, none of which has a `goog.provide`.

**Comment by gberaudo** [comment]
You are right, I tried moving the typedef back to attribution.js and it still compiles.
If we can figure out how to handle the documentation it would be a solution:

```js
var localOl = goog.require('ol');
/**
 * @typedef {string|Array.<string>|ol.Attribution|Array.<ol.Attribution>}
 * @api
 */
localOl.AttributionLike;
```

**Comment by probins** [comment]
There's a section in the [NTI docs](https://github.com/google/closure-compiler/wiki/Using-NTI-(new-type-inference)#typedefs-are-not-namespaces) 'Typedefs are not namespaces'. As they're not namespaces, they don't need a `provide` or a `require`.

The problem with your solution is that it adds code where code isn't needed. This would complicate the changeover to ES6. These things are only needed for the compiler's type checking and for the docs. So I would prefer the original idea of moving them out into a separate file, just without any `goog.provide`. Similarly, where they are referred to in the docs, they do not need a `goog.require`.

**Comment by probins** [comment]
yeah, on the docs, it looks like a typedef is only generated in `info.json` if there's a `goog.provide`. As not all typedefs have a `goog.provide`, this seems to be a hole in the current system. This needs changing so typedefs are handled separately.

**Comment by probins** [comment]
to be precise, a search tells me there are currently 73 `@typedef` in 57 files; 45 are defined in my `info.json`. An alternative to changing `generate-info.js` to handle the typedef file (assuming all typedefs are in one file) would be to make sure each entry has a `goog.provide`. Although the compiler doesn't need them, it should make sure they're all added to `info.json`.

**Comment by gberaudo** [comment]
Added support for `typedefs.js` files and moved some typedefs.
Fixed exports.js generation (but no handling of inheritance yet).
Fixed? apidoc generation (but no handling of inheritance yet).

**Comment by probins** [comment]
I was going to suggest putting the typedef stuff in a separate issue/PR. I'm away at the moment, but should have time to do this next week.

I'd also suggest creating GitHub milestones for removal of Closure lib and conversion to modules, listing the various steps as task lists. ISTM this is a big change and better split over several issues. Using milestones makes it clearer to people where the project is heading.

**Comment by gberaudo** [comment]
@probins, I wanted to wait for the end of the experiments in order to have a clear view about what works and what doesn't work and what we should do. But if you want to anticipate, please go on; it will help communicating about the current state and ongoing work.

**Comment by probins** [comment]
> Tried with ol.events.EventTarget: it did not work.

Not sure I understand this; `ol.events.EventTarget` isn't an enum.

**Comment by probins** [comment]
`ol.Collection.prototype.updateLength_` is a private method; private methods surely shouldn't be exported?

**Comment by gberaudo** [comment]
> Not sure I understand this; ol.events.EventTarget isn't an enum.

Good remark, I will have a new look to it.

> ol.Collection.prototype.updateLength_ is a private method; private methods surely shouldn't be exported?

It is not directly exported; only the constructor is exported using the `exports` variable`.

**Comment by probins** [comment]
If `ol.CollectionProperty` is compiling ok, I assume this is because of `LegacyNamespace`. In a pure/strict module this won't work because `ol` will be undefined. This isn't used outside the module, so could be simply renamed: `var CollectionProperty`. But this is then inconsistent with, say, `ol.ViewProperty`, which _is_ used outside the `View` module, albeit only in a very minor way, by the OverviewMap control. It too could be renamed, and then exported, for example, as `View#ViewProperty`. This would not affect those who use the api, but might affect those who compile code with the library, if they use the type in their code.

On the other hand, if `ViewProperty` is moved into its own module, then `CollectionProperty` should be too, even though doing so adds overhead of additional files for no benefit other than consistency.

Not sure what the best way of handling these enums is.

**Comment by probins** [comment]
I don't think [the change to closure-util for `typedefs.js`](https://github.com/openlayers/closure-util/commit/144620fd8703894a456a05cadc1524a0594d9464) is needed. To quote `tasks/readme.md`: "If you want to force the inclusion of files that don't use goog.require or goog.provide, you can use the js property of the compile object." So, if you just add `"js": ["src/ol/typedefs.js"],` to the compile object of `config/ol.json` and other build configs, the compiler finds the typedefs and compiles without problem. Alternatively, it could be hard-coded [here](https://github.com/openlayers/ol3/blob/master/tasks/build.js#L209).

**Comment by gberaudo** [comment]
Thanks @probins for double checking.

Enums have different uses in the library so I would not be suprised if we end with different ways to handle them. For example, there are goog.provided enums and enums that are not.

- Property enums
After discussion with @fredj, it appears these enums are mostly internal enums and it would make sense to keep them as private as possible, eventually only accessible in one file, like `CollectionProperty`.

- Other enums
Other enums like `ol.proj.Units` in proj.js are useful from the outside and should be exported. This particular enum is even marked `@api stable`.

I went on with enum experimentations. Having enums in the same file as a module for a namespace works though it requires a small change in applications compiled with closure (see https://github.com/gberaudo/ol3/commit/26cc3fa80770d4fde559810cbceec49c14f37cb0).

- Typedef files
I was unaware of the `compile.js` trick; I will try it. An advantage of having the change in `closure-util` is that it would work immediately for every project using `closure-util`, but maybe it would be too much magics.

**Comment by probins** [comment]
> it appears these enums are mostly internal enums and it would make sense to keep them as private as possible

I agree. In general, it doesn't make sense having a separate module just for an enum. Even when they need to be exported, it would be better to keep them where they are and rename them (`import Units from proj`). We could create a duplicate in a separate `ol.proj.Units` module for the time being purely for bc, and mark it as deprecated, encouraging people to change their code to the new non-namespace version. But judging from your comment in https://github.com/gberaudo/ol3/commit/26cc3fa80770d4fde559810cbceec49c14f37cb0 it looks like that might not work anyway.

**Comment by gberaudo** [comment]
Indeed, `ol.proj.Units` is a namespace enum. Based on my experiments we need to keep this enum in its namespace file.

**Comment by probins** [comment]
> I wanted to wait for the end of the experiments in order to have a clear view about what works and what doesn't work and what we should do

agreed. But I think it's now clear what to do with typedefs. As I said at the beginning of this issue, assuming we want to convert from globals to modules, this has to be done, and is independent of whether we go with `goog.module` or not. The same applies to the enums, but I'm not so sure atm what the solution is with them. As you say, it may make sense to deal with these in different ways. So for a start I'll set up a separate issue for the typedefs, and then work on a PR for them. Most of these global-to-module conversion tasks aren't particularly difficult, but there's a lot of rather tedious cut/paste and find/replace, so I might as well make a start on them while I have some time.

**Comment by gberaudo** [comment]
> agreed. But I think it's now clear what to do with typedefs. As I said at the beginning of this issue, assuming 

+1, I agree this task could already be started.

**Comment by gberaudo** [comment]
>Typedef files I was unaware of the compile.js trick; I will try it.

Ok, adding the typedef to `ol.conf` works. Great!

**Comment by gberaudo** [comment]
I have done more experiments with enums: splitting them in their own file can lead to circular dependency issue. The best solution for namespace enums seams to keep them there. Updating https://github.com/openlayers/ol3/issues/5226#issuecomment-215126637.

**Comment by probins** [comment]
yes, I like the solution with Corner/Relationship; nice and simple.

**Comment by probins** [comment]
On the subject of renaming the files to correspond to the module name, I tried a simple test, and Git is cleverer than I thought.

1. change a file in a new branch
2. rename the file in another branch
3. merge 2
4. merge 1 works as a fast-forward merge, with no conflict

so renaming the files shouldn't create merge conflicts as I thought. I was thinking of doing these file renames as a separate PR, but I now think it would be better to rename them as part of the process of converting the individual files to modules.

**Comment by gberaudo** [comment]
OK, let's try this way then.
For information, I am in the process of cleaning up my branch.

**Comment by probins** [comment]
> I like the solution with Corner/Relationship

I think I've changed my mind about this :-) enums are not exported at the moment, and in principle I don't think they should be in the future either. Corner and Relationship are only used in 1 other file. It happens that `ol.extent` is used in those other files, so in these cases adding Corner/Relationship to `ol.extent` won't increase the dependencies. However, it still means that something not exported ATM is now exported. These are however just a minor optimization for the compiler, and ISTM it would be better to just replace these refs with the string literal, for example, replace `ol.extent.Corner.TOP_LEFT` with `'top-left'`. Where they are referred to as types in comments, they are the same as typedefs, and could be moved to a single file like typedefs, renaming as necessary to get round the namespace circular dependency issues.

As you say, most of the enums are probably just used internally, and in this case can stop where they are. I'll do some analysis when looking into #5418 and see how big an issue there really is with enums.

**Comment by gberaudo** [comment]
I improved the test framework I mentionned in https://github.com/openlayers/ol3/issues/5623#issuecomment-234241969.

A call to `make test-doclets` in branch https://github.com/gberaudo/ol3/tree/goog_modules_squashed_rebased will check that the doclets produced by jsdoc stay unchanged and it could catch regressions if jsdoc behaviour changed in the future. This is essential for info/exports and the API documentation generation.

Hereafter is the current output:
```bash
gberaudo@wrk29:~/dev/ol3 (goog_modules_squashed_rebased $%)$ LC_ALL=C make test-doclets 
test_doclets/run.sh
test_doclets/compare.sh
Files test_doclets/inheritance/provide/result and test_doclets/inheritance/module/result differ
45,58d44
<     },
<     {
<       "name": "ol.Child#toOverride",
<       "kind": "function",
<       "stability": "experimental",
<       "inherited": true,
< 
<     },
<     {
<       "name": "ol.Child#onlyBase",
<       "kind": "function",
<       "stability": "experimental",
<       "inherited": true,
```

The inheritance test has to set of files:
- in the provide directory, a test case using goog.provide;
- in the module directory, a test case using the new goog.module mechanism.

Jsdoc is called on both directories then the result is normalized and compared.
Here we can see that the two methods `ol.Child#toOverride` and `ol.Child#onlyBase` are missing from the `goog.module` version.

@ahocevar, you know a lot about jsdoc and I see you did a lot of advanced tricks in the jsdoc api directory to handle inheritance. Do you have an idea of what I could do to overcome this?

From my experiments, this is the last blocking issue. It would be awesome to have this addressed and the work on `goog.module` and toward ES6 modules go forward.

**Comment by gberaudo** [comment]
The issue with inheritance is solved. It was caused by an incorrect `scope` property in the generated doclets. See https://github.com/gberaudo/ol3/commit/eb226525a764ccc794861577e98aa315c4979372.

Ths solution is really simple but tracking this has been time consuming.
All credits to FOSS4G Bonn for being so inspiring!

To me it was the last blocking issue for validating these experiments with `goog.module`.

Is everyone happy with the results?

Now I would like to start working on actual PRs.
@probins, @tschaub, @ahocevar, can I get a clear thumb up on this?

**Comment by probins** [comment]
@gberaudo is there a branch which includes some examples of the src changes you're proposing rebased to include all the recent work that's been done on typedefs, 1 file/1 require, say rebased to 3.18? What we discussed earlier in the year is a bit out of date now. It would be interesting to have a summary of what exactly needs to be done.

**Comment by gberaudo** [comment]
Yes @probins, the above commit is taken from https://github.com/gberaudo/ol3/tree/goog_modules_squashed_rebased which I rebased onto master yesterday.

**Comment by probins** [comment]
a couple of quick comments:
* this doesn't seem to be a complete rebase - see the 2 examples, for instance, which presumably should not be changed back to `goog.inherits`
* it would be good to have an example of import/require, perhaps also change `ol.TileRange` and import it into Attribution
* I think it would help with conversion to ES6 if there were only 1 `exports` statement/assignment per file, and follow that pattern throughout. In the `Attribution` case:
``` js
var Attribution = function...
Attribution.prototype ...
exports = Attribution;
```
which can then easily be converted to `export default Attribution` (I think ES6 should also have one consistent pattern, such as always default exports).

**Comment by ahocevar** [comment]
Would it make sense to pause this effort until #5794 is complete?

**Comment by gberaudo** [comment]
@probins, thanks for the feedback.
- it is a complete rebase, though not intended for merging but for validating the principle;
- yes I removed several experiments which where conflicting with recent changes in OL3;
  an interesting snippet is test_doclets/inheritance/module/*.js;
- would be great indeed, I need to check if it does not break the jsdoc plugin.

@ahocevar, I agree it would be ideal to start pushing PRs about `goog.module` after structural changes like #5794 are merged.

**Comment by probins** [comment]
> Would it make sense to pause this effort until #5794 is complete?

yes, a lot of this is dealing with enums, another of the tasks which had to be done anyway if converting to modules, so doing the work in #5794 means it doesn't have to be dealt with here.

**Comment by gberaudo** [comment]
There is also some work to be done on the ESLint rules.
The missing requires / valid requires ... does not handle assignements like `var something = goog.require('ol.something');`. In addition, they don't know about `goog.module` and the legacy declaration.
I have started to look into it.

**Comment by probins** [comment]
there's also an issue with the mapping of module names with filenames. goog modules have a name/identifier, but ES6 modules are anonymous and imported from a filename/url, so to convert `goog.require` to `import` you have to convert names to filenames/urls. Attribution has 3 dependencies: `ol.TileRange`, `ol.math` and `ol.tilegrid`. `ol.math` is in `ol/math.js`, so to convert you change '.' to '/' and append '.js'. `ol.TileRange` is in `ol/tilerange.js`, so you need to also convert names to lower-case. `ol.tilegrid` is in `ol/tilegrid/index.js`; how does a conversion script identify these and know it has to additionally append '/index' before the '.js'?

**Comment by tschaub** [comment]
The rules are intentionally designed to work toward one provide per module (#5679).

**Comment by tschaub** [comment]
I think it makes sense to stay focused on the goal of increasing usability and interoperability.  The work around #5679 is about that (interoperability with CJS tools).  It feels like a bit of a distraction to me to focus on things like goog.module and converting that to ES modules.

I'm now confident in the potential for transpiling to CJS modules after #5679.  This will allow for huge interoperability gains in the 3.x release series.

I'm pretty sure that authoring in ES modules will require a 4.0 release.  And it is not clear to me where goog.module fits in.

**Comment by probins** [comment]
as I said in https://github.com/openlayers/ol3/issues/5679#issuecomment-237486360, `goog.module` would simplify the transpile. It enables you to convert the code from global namespaces to a standard module format whilst remaining compatible with the current compile/build process. So you wouldn't need to transpile `ol.foo.Bar` to `$ol$foo$Bar` or whatever, as `ol.foo.Bar` would no longer be in the code. To transpile to CJS, you would just get rid of the `goog.module` line, and change the `goog.require` line to a CJS `require`. Similarly for transpile to ES6, which could be used as the basis for a new OL4. (OTOH, if OL4 is more of a rewrite, using ES6 classes for example, that might not be such a big advantage.)

Personally, I just want modules in a form I can load in the browser. If the decision is to go with `goog.module`, I'll convert that. But I can just as easily use CJS with `$ol$foo$Bar`.

**Comment by gberaudo** [comment]
As @probins said, using `goog.module` allows structuring the code cleanly in OpenLayers 3 and easily transpiling to CJS or ES6 modules.

It is also the official transition path proposed by Google from its proprietary dependency system to standard ES6 modules.
It allows people using the closure compiler to transition their applications gradually.

Having a path for the transition is appealing to us at Camptocamp (@sbrunner, @fredj) and we are willing to work on it.

We have overcome all the difficulties we encountered during the experiments with  `goog.module`.
@tschaub have we missed something? Is there something wrong with `goog.module`?

**Comment by probins** [comment]
> Would it make sense to pause this effort until #5794 is complete?

it might have been simpler to have changed to `goog.module` first. Then the linter checks for missing/unused `goog.require`s etc wouldn't be needed; any issues would be picked up by the standard linter/compiler checks for unused/undeclared vars.

**Comment by gberaudo** [comment]
@probins see https://github.com/gberaudo/ol3/commits/goog_modules_squashed_rebased_local_names (first commit) for a test of using a single assignement to exports. It looks like bringing more issues than solving.

**Comment by probins** [comment]
The issues there are with JSDoc, not the compiler, right?

It's not a major problem, as you should be able to convert to ES6 by changing `exports =` to `var exports =`, and then have at the end `export default exports`. More important is to use the same pattern throughout.

**Comment by gberaudo** [comment]
Yes these are the issues with JSDoc. Your solution with `export default exports` looks great to me.

**Comment by probins** [comment]
@gberaudo I've been looking at @tschaub's [codemod](https://github.com/openlayers/ol3/compare/master...tschaub:mod) and have pushed the results to https://github.com/probins/olcjs (for comparison, the same converted to ES6 is at https://github.com/probins/oles6). I commented before that it should be straightforward to transform `goog.module`s to CJS, but I'm thinking that the reverse should also be true. If this is so, if you transform the CJS version to `goog.module` (a sort of back-port), then compiling the output should produce the same results as using the existing `src/ol` files. Like to try this out, and see if you agree?

**Comment by gberaudo** [comment]
@probins thank you for your investigations.

What looks easiest to me is to:
- have OpenLayers code use `goog.module` together with `goog.module.declareLegacyNamespace()`;
- have a converter from `goog.module` to ES6 modules;
- have a converter from ES6 modules to CommonJS.

We would need to implement the converter from `goog.module` to ES6 modules.
I guess there are already many projects for converting ES6 modules to CommonJS.

When OpenLayers decides to abandon `goog.require/goog.provide/goog.module` it will be transparent for CommonJS users and ES6 module users.

**Comment by tschaub** [comment]
Can you describe what the goal of using `goog.module` is?

Personally, I don't think that "make the code a bit more like ES6 modules" is a valid reason to increase our dependency on the Closure Library.

I do think things like "increase library adoption" are valid goals.  And this is the motivation for moving away from the Closure Library and the hard dependency on the Closure Compiler.

**Comment by tschaub** [comment]
Also, it is straightforward to convert the current code base to CommonJS modules or ES6 modules.  There is no need to use `goog.module` as part of this.

**Comment by bartvde** [comment]
I agree with @tschaub on this, I also don't see the need for goog.module

**Comment by gberaudo** [comment]
Here is my question: our customers code is written with `goog.require` and `goog.provide`. How do you see the transition to ES6 modules for them?

At Camptocamp we have thoroughly thought about this and we think `goog.module` is a great way to achieve this. As explained in the description of this proposition and the links, `goog.module` allows mixing the current and new code style at the same time, allowing a gradual transition:
- current style
```js
goog.require('ol.Map');
```
- ES6-like style
```js
var Map = goog.require('ol.Map');
```

We think `goog.module` is a great opportunity for our customers.

As explained in https://github.com/openlayers/closure-util/pull/267#issuecomment-267056095 it will not  `increase our dependency on the Closure Library`.

**Comment by ahocevar** [comment]
* CommonJS style:
```js
var Map = require('ol/Map');
```
I don't see how this would be any harder for your customers than the ES6-like style (which does not look very ES6-like to me) you showed above.

**Comment by probins** [comment]
CJS/ES6 modules are not compatible with ol3; `goog.module` ones are. `goog.require(...)` establishes a dependency; it's not an import. `var xxx = goog.require(...)` is an import and is the exact equivalent of `import xxx from ...`.

@tschaub - a question. The code output by your codemod, as shown in https://github.com/tschaub/ol3/blob/e652aa81ca7e82c6afb21afbe9433f6b031837fa/package/readme.md, cannot be used by existing apps. Incompatible code means major-version change, so are you proposing this as ol4?

**Comment by tschaub** [comment]
> Rationnal: gradual and non-breaking step towards ES6

@gberaudo - it would be useful if you could add a bit more detail to what you think using `goog.module()` achieves.

For example, is there something specific that library users can do after your proposed change that they cannot do now?

If this proposed change doesn't make it possible to do anything more with the library, then can you describe what you think the next step would be?  If the next step is "convert the library to ES6 modules," then why is this intermediate step useful?

**Comment by tschaub** [comment]
> @tschaub - a question. The code output by your codemod, as shown in https://github.com/tschaub/ol3/blob/e652aa81ca7e82c6afb21afbe9433f6b031837fa/package/readme.md, cannot be used by existing apps. Incompatible code means major-version change, so are you proposing this as ol4?

This is the wrong place for this discussion.  But, since you asked.  During our [release process](https://github.com/openlayers/openlayers/wiki/Release-Procedure) we publish the `openlayers` package.  The changes in #6302 would introduce another step in the release process - specifically, publishing an `ol` package.  Developers can continue using `closure-util` with the `openlayers` package.  So the short answer to your question is "no" (I am not proposing anything that breaks backwards compatibility, and this is not about a 4.x version bump).

**Comment by tschaub** [comment]
Instead of `goog.module()`, we are going to use ES modules.
